File: python/sel/builtins/null.py
from ..registry import INF, define
from ..value import Value


def _is_null(args, ctx):
    return Value.bool(args.val(0).is_null())


def _is_not_null(args, ctx):
    return Value.bool(not args.val(0).is_null())


def _coalesce(args, ctx):
    for i in range(args.count()):
        v = args.val(i)
        if not v.is_null():
            return v
    return Value.null()


def _get(args, ctx):
    target = args.val(0)
    key = args.text(1)
    if not target.is_null() and target.has(key):
        val = target.get(key)
        if val is not None:
            return val
    if args.count() > 2:
        return args.val(2)
    return Value.null()


def _path(args, ctx):
    target = args.val(0)
    path_str = args.text(1)
    if path_str == '':
        return target
    segments = path_str.split('.')
    cur = target
    for seg in segments:
        if cur.is_null() or not cur.has(seg):
            if args.count() > 2:
                return args.val(2)
            return Value.null()
        cur = cur.get(seg)
        if cur is None:
            if args.count() > 2:
                return args.val(2)
            return Value.null()
    return cur


def _is_blank(args, ctx):
    return Value.bool(args.val(0).is_vacuous())


def _is_present(args, ctx):
    return Value.bool(not args.val(0).is_vacuous())


define('IS_NULL', 1, 1, fn=_is_null)
define('IS_NOT_NULL', 1, 1, fn=_is_not_null)
define('COALESCE', 1, INF, lazy=True, fn=_coalesce)
define('GET', 2, 3, lazy=True, fn=_get)
define('PATH', 2, 3, lazy=True, fn=_path)
define('IS_BLANK', 1, 1, fn=_is_blank)
define('IS_PRESENT', 1, 1, fn=_is_present)

File: js/src/builtins/null.mjs
import { Value } from '../value.mjs';
import { define } from '../registry.mjs';

define({
  name: 'IS_NULL', min: 1, max: 1,
  fn: (args) => Value.bool(args.val(0).isNull()),
});

define({
  name: 'IS_NOT_NULL', min: 1, max: 1,
  fn: (args) => Value.bool(!args.val(0).isNull()),
});

define({
  name: 'COALESCE', min: 1, max: Infinity, lazy: true,
  fn: (args) => {
    const n = args.count();
    for (let i = 0; i < n; i++) {
      const v = args.val(i);
      if (!v.isNull()) return v;
    }
    return Value.null();
  },
});

define({
  name: 'GET', min: 2, max: 3, lazy: true,
  fn: (args) => {
    const target = args.val(0);
    const key = args.text(1);
    if (!target.isNull() && target.has(key)) {
      const v = target.get(key);
      if (v !== undefined) return v;
    }
    if (args.count() > 2) return args.val(2);
    return Value.null();
  },
});

define({
  name: 'PATH', min: 2, max: 3, lazy: true,
  fn: (args) => {
    const target = args.val(0);
    const pathStr = args.text(1);
    if (pathStr === '') return target;
    const segments = pathStr.split('.');
    let cur = target;
    for (const seg of segments) {
      if (cur.isNull() || !cur.has(seg)) {
        if (args.count() > 2) return args.val(2);
        return Value.null();
      }
      cur = cur.get(seg);
      if (cur === undefined) {
        if (args.count() > 2) return args.val(2);
        return Value.null();
      }
    }
    return cur;
  },
});

define({
  name: 'IS_BLANK', min: 1, max: 1,
  fn: (args) => Value.bool(args.val(0).isVacuous()),
});

define({
  name: 'IS_PRESENT', min: 1, max: 1,
  fn: (args) => Value.bool(!args.val(0).isVacuous()),
});
